File: src/main/java/io/github/takenoha/towerdefense/config/TowerProfile.java
package io.github.takenoha.towerdefense.config;

/** Configurable combat profile for one of the specialist tower types. */
public record TowerProfile(
        int damage,
        double range,
        int attackIntervalTicks,
        double areaRadius,
        double slowPercent,
        int slowDurationTicks,
        int chainCount,
        double chainRadius,
        double supportRadius,
        double supportDamageMultiplier,
        double supportSpeedMultiplier,
        double supportRangeMultiplier,
        int supportStackLimit,
        int burnDurationTicks) {

    public static TowerProfile frostDefaults() {
        return new TowerProfile(2, 12.0d, 30, 0.0d, 0.35d, 50, 0, 0.0d,
                0.0d, 1.0d, 1.0d, 1.0d, 0, 0);
    }

    public static TowerProfile lightningDefaults() {
        return new TowerProfile(6, 18.0d, 35, 0.0d, 0.0d, 0, 3, 5.0d,
                0.0d, 1.0d, 1.0d, 1.0d, 0, 0);
    }

    public static TowerProfile supportDefaults() {
        return new TowerProfile(1, 10.0d, 40, 0.0d, 0.0d, 0, 0, 0.0d,
                8.0d, 1.25d, 0.80d, 1.15d, 2, 0);
    }

    public static TowerProfile sniperDefaults() {
        return new TowerProfile(18, 32.0d, 60, 0.0d, 0.0d, 0, 0, 0.0d,
                0.0d, 1.0d, 1.0d, 1.0d, 0, 0);
    }

    public static TowerProfile flameDefaults() {
        return new TowerProfile(3, 15.0d, 25, 3.0d, 0.0d, 0, 0, 0.0d,
                0.0d, 1.0d, 1.0d, 1.0d, 0, 80);
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/config/TowerProfile.kt
package io.github.takenoha.towerdefense.config

import kotlin.jvm.JvmRecord

/** Configurable combat profile for one of the specialist tower types. */
@JvmRecord
data class TowerProfile(
    val damage: Int,
    val range: Double,
    val attackIntervalTicks: Int,
    val areaRadius: Double,
    val slowPercent: Double,
    val slowDurationTicks: Int,
    val chainCount: Int,
    val chainRadius: Double,
    val supportRadius: Double,
    val supportDamageMultiplier: Double,
    val supportSpeedMultiplier: Double,
    val supportRangeMultiplier: Double,
    val supportStackLimit: Int,
    val burnDurationTicks: Int,
) {
    companion object {
        @JvmStatic
        fun frostDefaults(): TowerProfile = TowerProfile(
            2, 12.0, 30, 0.0, 0.35, 50, 0, 0.0,
            0.0, 1.0, 1.0, 1.0, 0, 0,
        )

        @JvmStatic
        fun lightningDefaults(): TowerProfile = TowerProfile(
            6, 18.0, 35, 0.0, 0.0, 0, 3, 5.0,
            0.0, 1.0, 1.0, 1.0, 0, 0,
        )

        @JvmStatic
        fun supportDefaults(): TowerProfile = TowerProfile(
            1, 10.0, 40, 0.0, 0.0, 0, 0, 0.0,
            8.0, 1.25, 0.80, 1.15, 2, 0,
        )

        @JvmStatic
        fun sniperDefaults(): TowerProfile = TowerProfile(
            18, 32.0, 60, 0.0, 0.0, 0, 0, 0.0,
            0.0, 1.0, 1.0, 1.0, 0, 0,
        )

        @JvmStatic
        fun flameDefaults(): TowerProfile = TowerProfile(
            3, 15.0, 25, 3.0, 0.0, 0, 0, 0.0,
            0.0, 1.0, 1.0, 1.0, 0, 80,
        )
    }
}
